Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp

```diff
@@ -261,6 +263,20 @@ const std::optional<ScreenOrientationLockType> ApplicationManifestParser::parseO
     return std::nullopt;
 }
 
+String ApplicationManifestParser::parseLang(const JSON::Object& manifest)
+{
+    auto lang = parseGenericString(manifest, "lang"_s);
+    if (lang.isEmpty())
+        return { };
+
+    if (!JSC::isStructurallyValidLanguageTag(lang)) {
+        logDeveloperWarning(makeString('"', lang, "\" is not a structually valid language tag."_s));
+        return { };
+    }
+
+    return JSC::canonicalizeUnicodeLocaleID(lang.utf8());
```

Comment:
Converting to UTF-8 on the way in is a peculiar step in canonicalizing this ID, and even an unnecessarily inefficient one since a structurally valid language tag is guaranteed not to have non-ASCII characters and likely is almost always already stored as an 8-bit string, while the `utf8` function always returns newly allocated memory than then has to be deallocated. Building this inefficient idiom into the interface between JavaScriptCore and WebCore isn’t great, even for something trivial like this. The conversion step is required only because the existing function from `IntlObject.h` happens to take a `CString`. I think we should have a function that takes a `String` or `StringView` that does the same operation that we export from `IntlObject.h`, even if for now it has creates a `CString` internally for reasons of expediency; it would be easy to write a version that doesn’t unnecessarily allocate and deallocate memory for an extra copy of the string in the common case, though.